iceoryx_dust/include/iceoryx_dust/internal/singleton.inl

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_DUST_SINGLETON_INL
+#define IOX_DUST_SINGLETON_INL
+
+#include "iceoryx_dust/singleton.hpp"
+
+namespace iox
+{
+
+template <typename T>
+bool Singleton<T>::isInitialized()
+{
+    return ptr().load(std::memory_order_relaxed) != nullptr;
+}
+
+template <typename T>
+template <typename... Args>
+T& Singleton<T>::init(Args&&... args)
+{
+    std::lock_guard<std::mutex> g(lock());
+    if (!isInitialized())
+    {
+        // initialized by this call
+        return *initialize(std::forward<Args>(args)...);
+    }
+    // initialized before by some other call
+    return *ptr().load(std::memory_order_acquire);
+}
+
+template <typename T>
+void Singleton<T>::destroy()
+{
+    std::lock_guard<std::mutex> g(lock());
+    auto p = ptr().load(std::memory_order_acquire);
+    if (p)
+    {
+        p->~T();
+        ptr().store(nullptr);
+    }
+}
+
+template <typename T>
+T& Singleton<T>::instance()
+{
+    // need to sync the memory at *p as well
+    auto p = ptr().load(std::memory_order_acquire);
```

Comment:
@MatthiasKillat just to give some context to my previous statement why I think this construct is quite complex and it will cost me too much time to review it now.
This line breaks the singleton since one thread can call `instance()` and obtain the pointer while another thread can easily call `destroy` followed by `init`. This is totally valid with the pubic API but leaves the first thread with a dangling reference. One cannot know if someone calls `destroy`, therefore `instance` is inherently unsafe to use.

Reply:
I know. This is intentional and cannot easily be guarded against (without huge performance hits, if even possible). A real Meyers Singleton has the same issue, just more subtle (another static can hold a reference to it and use this reference after the dtor was called).

The main purpose is to call destroy only during some shutdown function ad not defer destruction to some point after main.

Reply:
@MatthiasKillat right. The main issue is that the API can easily be misused but we talked already about a possible alternative.